src/QuickCheckVEngine/RVFI_DII/RVFI.hs

```diff
@@ -73,175 +84,404 @@ type RV_RegIdx = Word8
+      rvfi_mem_wmask = 0,
+      rvfi_mem_rdata = 0,
+      rvfi_mem_wdata = 0
+    }
+
+hexStr :: BS.ByteString -> String
+hexStr msg = show (toLazyByteString (lazyByteStringHex msg))
+
+type ConnectionInfo = (String, Int)
+
+connectionDebugMessage :: Int -> ConnectionInfo -> String -> IO ()
+connectionDebugMessage minVerbosity (name, verbosity) msg = do
+  when (verbosity >= minVerbosity) $
+    putStrLn ("\t" ++ name ++ ": " ++ msg)
+
+errorWithContext :: String -> String -> IO ()
+errorWithContext name msg = do
+  error (name ++ ": " ++ msg)
+
+connectionError :: ConnectionInfo -> String -> IO ()
+connectionError (name, _) msg = do
+  errorWithContext name msg
+
+rvfiCheckMagicBytes :: BS.ByteString -> String -> ConnectionInfo -> IO ()
+rvfiCheckMagicBytes magicBytes expected conn = do
+  let expBytes = C8.pack expected
+  connectionDebugMessage 3 conn ("read header magic bytes: " ++ show magicBytes)
+  when (magicBytes /= expBytes) $
+    connectionError conn ("received invalid data packet: got magic=" ++ show magicBytes ++ " but expected " ++ show expBytes)
+  return ()
+
+rvfiReadDataPacketWithMagic :: (Int64 -> IO BS.ByteString, String, Int) -> Int64 -> String -> IO BS.ByteString
+rvfiReadDataPacketWithMagic (reader, name, verbosity) size expectedMagic = do
+  when (size < 8) $
+    errorWithContext name ("Invalid packet size:" ++ show size)
+  msg <- reader size
+  connectionDebugMessage 3 (name, verbosity) ("read packet: " ++ hexStr msg)
+  let (magic, bytes) = BS.splitAt 8 msg
+  rvfiCheckMagicBytes magic expectedMagic (name, verbosity)
+  return bytes
+
+rvfiReadV2Response :: (Int64 -> IO BS.ByteString, String, Int) -> IO RVFI_Packet
+rvfiReadV2Response (reader, name, verbosity) = do
+  let connInfo = (name, verbosity)
+  connectionDebugMessage 3 connInfo "reading V2 packet..."
+  headerBytes <- rvfiReadDataPacketWithMagic (reader, name, verbosity) 64 "trace-v2"
+  let (traceSizeBytes, payloadBytes) = BS.splitAt 8 headerBytes
+  let traceSize = runGet getWord64le traceSizeBytes
+  connectionDebugMessage 3 connInfo ("trace-v2 common payload bytes: " ++ hexStr payloadBytes)
+  -- Ensure that we read all bytes in the packet
+  let (basicData, availableFeatures) = runGet (isolate 48 rvfiDecodeV2Header) payloadBytes
+  connectionDebugMessage 3 connInfo ("features: " ++ (printf "0x%016x" availableFeatures))
+  (intData, rf1, numBytes1) <- rvfiMaybeReadIntData (reader, name, verbosity) availableFeatures
+  (memData, rf2, numBytes2) <- rvfiMaybeReadMemData (reader, name, verbosity) rf1
+  when (rf2 /= 0) $
+    errorWithContext name ("Remaining unknown feature bits set: " ++ show rf2)
+  let remainingBytes = (fromIntegral traceSize) - 64 - numBytes1 - numBytes2
+  when (remainingBytes /= 0) $
+    errorWithContext name ("Did not read all bytes of V2 trace packet: " ++ show remainingBytes ++ " remaining")
+  return $ basicData {rvfi_int_data = intData, rvfi_mem_data = memData}
+
+rvfiDecodeV2Header :: Get (RVFI_Packet, Word64)
```

Comment:
I think for that one it would be easier to follow with a "type as documentation" approach (as in have `type AvailableRVFIFeatures = Word64` and then use that everywhere instead). What do you think?

Reply:
Make it `RVFIFeatures` as you often end up taking the intersection of requested and available to compute the negotiated set.